r/src/altrep.cpp

```diff
@@ -0,0 +1,180 @@
+            array->type()->ToString().c_str(), array->length(), array.get());
+    return TRUE;
+  }
+
+  static std::shared_ptr<Array>& Get(SEXP vec) {
+    return *cpp11::external_pointer<std::shared_ptr<Array>>(R_altrep_data1(vec));
+  }
+
+  static R_xlen_t Length(SEXP vec) { return Get(vec)->length(); }
+
+  static const void* Dataptr_or_null(SEXP vec) {
+    auto& array = Get(vec);
+
+    int size = array->type_id() == Type::INT32 ? sizeof(int) : sizeof(double);
+    return reinterpret_cast<const void*>(array->data()->buffers[1]->data() +
+                                         size * array->offset());
+  }
+
+  static SEXP Duplicate(SEXP vec, Rboolean) {
+    auto& array = Get(vec);
+    auto size = array->length();
+
+    bool int_array = array->type_id() == Type::INT32;
+
+    SEXP copy = PROTECT(Rf_allocVector(int_array ? INTSXP : REALSXP, array->length()));
+
+    memcpy(DATAPTR(copy), Dataptr_or_null(vec),
+           int_array ? (size * sizeof(int)) : (size * sizeof(double)));
+
+    UNPROTECT(1);
+    return copy;
+  }
+
+  static void* Dataptr(SEXP vec, Rboolean writeable) {
+    return const_cast<void*>(Dataptr_or_null(vec));
+  }
+
+  // by definition, there are no NA
+  static int No_NA(SEXP vec) { return 1; }
+
+  static void Init(R_altrep_class_t class_t, DllInfo* dll) {
+    // altrep
+    R_set_altrep_Length_method(class_t, array_nonull::Length);
+    R_set_altrep_Inspect_method(class_t, array_nonull::Inspect);
+    R_set_altrep_Duplicate_method(class_t, array_nonull::Duplicate);
+
+    // altvec
+    R_set_altvec_Dataptr_method(class_t, array_nonull::Dataptr);
+    R_set_altvec_Dataptr_or_null_method(class_t, array_nonull::Dataptr_or_null);
+  }
+};
+
+struct array_nonull_dbl_vector {
+  static R_altrep_class_t class_t;
+
+  static void Init(DllInfo* dll) {
+    class_t = R_make_altreal_class("array_nonull_dbl_vector", "arrow", dll);
+    array_nonull::Init(class_t, dll);
+    R_set_altreal_No_NA_method(class_t, array_nonull::No_NA);
+  }
+};
+
+struct array_nonull_int_vector {
+  static R_altrep_class_t class_t;
+
+  static void Init(DllInfo* dll) {
+    class_t = R_make_altinteger_class("array_nonull_int_vector", "arrow", dll);
+    array_nonull::Init(class_t, dll);
+    R_set_altinteger_No_NA_method(class_t, array_nonull::No_NA);
+  }
+};
+
+R_altrep_class_t array_nonull_int_vector::class_t;
+R_altrep_class_t array_nonull_dbl_vector::class_t;
+
+void Init_Altrep_classes(DllInfo* dll) {
+  array_nonull_dbl_vector::Init(dll);
+  array_nonull_int_vector::Init(dll);
+}
+
+SEXP Make_array_nonull_dbl_vector(const std::shared_ptr<Array>& array) {
+  return array_nonull::Make(array_nonull_dbl_vector::class_t, array);
+}
+
+SEXP Make_array_nonull_int_vector(const std::shared_ptr<Array>& array) {
+  return array_nonull::Make(array_nonull_int_vector::class_t, array);
+}
+
+}  // namespace r
+}  // namespace arrow
+
+// TODO: when arrow depends on R 3.5 we can eliminate this
```

Comment:
What does this mean? Does Arrow still support R versions before 3.5? @nealrichardson